src/arkode/arkode_lsrkstep.c

```diff
@@ -1472,59 +1854,289 @@ int lsrkStep_TakeStepSSP104(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPt
                ark_mem->ycur, ark_mem->ycur);
   for (int j = 6; j <= 9; j++)
   {
+#if SUNDIALS_LOGGING_LEVEL >= SUNDIALS_LOGGING_DEBUG
+    SUNLogger_QueueMsg(ARK_LOGGER, SUN_LOGLEVEL_DEBUG,
+                       "ARKODE::lsrkStep_TakeStepSSP104", "start-stage",
+                       "step = %li, stage = %i, h = %" RSYM ", tcur = %" RSYM,
+                       ark_mem->nst, j, ark_mem->h,
+                       ark_mem->tcur +
+                         ((sunrealtype)j - FOUR) * onesixth * ark_mem->h);
+#endif
+
     retval = step_mem->fe(ark_mem->tcur +
                             ((sunrealtype)j - FOUR) * onesixth * ark_mem->h,
-                          ark_mem->ycur, ark_mem->fn, ark_mem->user_data);
+                          ark_mem->ycur, step_mem->Fe, ark_mem->user_data);
     step_mem->nfe++;
-    if (retval != ARK_SUCCESS) { return (ARK_RHSFUNC_FAIL); }
+    if (retval < 0) { return ARK_RHSFUNC_FAIL; }
+    if (retval > 0) { return RHSFUNC_RECVR; }
+
+#ifdef SUNDIALS_LOGGING_EXTRA_DEBUG
+    SUNLogger_QueueMsg(ARK_LOGGER, SUN_LOGLEVEL_DEBUG,
+                       "ARKODE::lsrkStep_TakeStepSSP104", "stage RHS",
+                       "F_%i(:) =", j);
+    N_VPrintFile(step_mem->Fe, ARK_LOGGER->debug_fp);
+#endif
 
-    N_VLinearSum(ONE, ark_mem->ycur, onesixth * ark_mem->h, ark_mem->fn,
+    N_VLinearSum(ONE, ark_mem->ycur, onesixth * ark_mem->h, step_mem->Fe,
                  ark_mem->ycur);
 
-    if (j == 7)
+    if (j == 7 && !ark_mem->fixedstep)
     {
-      N_VLinearSum(ONE, ark_mem->tempv1, onefifth * ark_mem->h, ark_mem->fn,
+      N_VLinearSum(ONE, ark_mem->tempv1, onefifth * ark_mem->h, step_mem->Fe,
                    ark_mem->tempv1);
     }
-    if (j == 9)
+    if (j == 9 && !ark_mem->fixedstep)
     {
       N_VLinearSum(ONE, ark_mem->tempv1, SUN_RCONST(0.3) * ark_mem->h,
-                   ark_mem->fn, ark_mem->tempv1);
+                   step_mem->Fe, ark_mem->tempv1);
+    }
+
+    /* apply user-supplied stage postprocessing function (if supplied) */
+    if (ark_mem->ProcessStage != NULL)
```

Comment:
Post-processes embedding too

Reply:
I disagree.  The purpose of `ProcessStage` is to allow users to update any auxiliary data structures once the current stage has changed, so that they can be ready for a subsequent call to the user-supplied RHS function.  We should *never* call RHS functions on the embedded solution, nor is it ever considered to be a "current" solution.

Reply:
It is also used (at the user's own risk) to modify solution values e.g., to force small negative values to zero. Since this state is passed to an RHS evaluation users may wish to preform such actions on this state. In other steppers (e.g., ARKStep and ERKStep) we "effectively" apply the post-processing functions to the embedding because it is constructed from post processes stage RHS evaluations (and difference in embedding and method coefficients). My suggestion is to be consistent with what is done in the other steppers but maybe this only should be done before computing the difference in the solutions.

Reply:
I think I'm mixing up this instance (where the state is not passed to the RHS) with another case (where a state is passed to the RHS). Since this is not passed to the RHS, processing it is not needed, except maybe before computing the error estimate.

Reply:
I've gone back over my post-processing comments to clarify when they are not needed (on embeddings) and when they are (usually an initial stage computation before the for loop). This is one odd case, see the discussion [here](https://github.com/LLNL/sundials/pull/541#discussion_r1810061910).